pkg/io/video/scale.go

```diff
@@ -0,0 +1,175 @@
+package video
+
+import (
+	"errors"
+	"image"
+
+	"golang.org/x/image/draw"
+)
+
+var (
+	scalerTestAlgos = map[string]Scaler{
+		"NearestNeighbor": ScalerNearestNeighbor,
+	}
+	scalerBenchAlgos = map[string]Scaler{
+		"NearestNeighbor": ScalerNearestNeighbor,
+		"ApproxBiLinear":  ScalerApproxBiLinear,
+		"BiLinear":        ScalerBiLinear,
+	}
+)
+
+// Scaler represents scaling algorithm
+type Scaler draw.Scaler
+
+// List of scaling algorithms
+var (
+	ScalerNearestNeighbor = Scaler(draw.NearestNeighbor)
+	ScalerApproxBiLinear  = Scaler(draw.ApproxBiLinear)
+	ScalerBiLinear        = Scaler(draw.BiLinear)
+	ScalerCatmullRom      = Scaler(draw.CatmullRom)
+)
+
+var errUnsupportedImageType = errors.New("scaling: unsupported image type")
+
+// Scale returns video scaling transform.
+// Setting scaler=nil to use default scaler. (ScalerNearestNeighbor)
+// Negative width or height value will keep the aspect ratio of incoming image.
+//
+// Note: computation cost to scale YCbCr format is 10 times higher than RGB
+// due to the implementation in x/image/draw package.
+func Scale(width, height int, scaler Scaler) TransformFunc {
+	scalerCached := ScalerNearestNeighbor
+	if scaler != nil {
+		scalerCached = scaler
+	}
+	cacheScaler := func(dRect, sRect image.Rectangle) {
+		if kernel, ok := scaler.(interface {
+			NewScaler(int, int, int, int) draw.Scaler
+		}); ok {
+			scalerCached = kernel.NewScaler(dRect.Dx(), dRect.Dy(), sRect.Dx(), sRect.Dy())
+		}
+	}
```

Comment:
Rather than checking scaler interface inside of the func, it can be moved outside, and inside the function, it can just be

```
scalerCached = kernel.NewScaler(dRect.Dx(), dRect.Dy(), sRect.Dx(), sRect.Dy())
```